### PR TITLE
Allow users to add environment variables in the configuration file. #1282

### PR DESCRIPTION
**Pull request by leon0707**
**Proposed changes**:
I created the issue #1282. This pr is suppose to fix it.

- allow users to add environment variables in the configuration file. 

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog

### CODE REVIEW

**Comment by akelad** [comment]
Thanks for submitting this PR, I think this could be really useful! We'll give it a review as soon as possible

**Comment by wochinge** [comment]
@leon0707 This is an awesome idea! 

I try it out later today, just a few things
- if we can do something like `${VARIABLE}/remaining` -> `valueOfVariable/remaining` it might be nice to also support `prefix/${VARIABLE}` -> `prefix/valueOfVariable`
- some tests would be great
  - environment variable present (no prefix, prefix, no remaining path, remaining path)
  - environment variable not present

File: rasa_core/utils.py [hunk 1]
```diff
@@ -328,12 +328,39 @@ def construct_yaml_str(self, node):
...
+    def env_var_constructor(loader, node):
+        """Process environment variables found in the YAML."""
+        value = loader.construct_scalar(node)
+        envVar, remainingPath = env_var_pattern.match(value).groups()
+        return os.environ[envVar] + remainingPath
+
+    if six.PY2:
+        yaml.add_constructor(u'!env_var', env_var_constructor)
+    else:
+        yaml.constructor.SafeConstructor.add_constructor(
```
Comment (wochinge) [inline rasa_core/utils.py]:
```suggestion
        yaml.SafeConstructor.add_constructor(
```

File: rasa_core/utils.py [hunk 2]
```diff
@@ -328,12 +328,39 @@ def construct_yaml_str(self, node):
...
+    import os
+
+    # eg. ${USER_NAME}, ${PASSWORD}
+    env_var_pattern = re.compile(r"^\$\{(.*)\}(.*)$")
+    yaml.add_implicit_resolver('!env_var', env_var_pattern)
+
+    def env_var_constructor(loader, node):
+        """Process environment variables found in the YAML."""
+        value = loader.construct_scalar(node)
+        envVar, remainingPath = env_var_pattern.match(value).groups()
```
Comment (wochinge) [inline rasa_core/utils.py]:
```suggestion
        env_var, remaining_path = env_var_pattern.match(value).groups()
```

File: rasa_core/utils.py [hunk 3]
```diff
@@ -328,12 +328,39 @@ def construct_yaml_str(self, node):
...
+        return os.environ[envVar] + remainingPath
```
Comment (wochinge) [inline rasa_core/utils.py]:
```suggestion
        return os.environ[env_var] + remaining_path
```

File: rasa_core/utils.py [hunk 4]
```diff
@@ -328,12 +328,39 @@ def construct_yaml_str(self, node):
                                     construct_yaml_str)
 
 
+def process_env_var():
```
Comment (wochinge) [inline rasa_core/utils.py]:
```suggestion
def replace_environment_variables():
```

File: rasa_core/utils.py [hunk 5]
```diff
@@ -328,12 +328,39 @@ def construct_yaml_str(self, node):
...
+            u'!env_var', env_var_constructor)
+
+
 def read_yaml_file(filename):
     """Read contents of `filename` interpreting them as yaml."""
     return read_yaml_string(read_file(filename))
 
 
 def read_yaml_string(string):
+    process_env_var()
```
Comment (wochinge) [inline rasa_core/utils.py]:
```suggestion
    replace_environment_variables()
```

**Comment by wochinge** [comment]
So checked it out and played around a bit: Works like a charm 🚀 
I have a few suggestions on parameter naming, and the ones from the comment, but otherwise it's fine! Thanks for your contribution!

**Comment by leon0707** [comment]
Thank you @CLAassistant @wochinge I will update the names and add test cases.

**Comment by wochinge** [comment]
@leon0707 If you want so inspiration for the tests: Have a look at my pull request for NLU: https://github.com/RasaHQ/rasa_nlu/pull/1518

**Comment by leon0707** [comment]
@wochinge I added a couple test cases. `prefix_${XXX}`, `prefix_${XXX}_postfix` and `${XXX}_postfix` are all supported.

File: tests/test_utils.py [hunk 6]
```diff
@@ -118,3 +119,112 @@ def test_endpoint_config():
...
+         host: http://localhost:27017
+         db: rasa
+         user: ${USER_NAME}
+         password: ${PASSWORD}
+         collection: conversations
+    """
+    try:
+        r = utils.read_yaml_string(config_with_env_var)
+    except Exception as e:
+        assert isinstance(e, KeyError) and e.args[0] == 'PASSWORD'
```
Comment (wochinge) [inline tests/test_utils.py]: please use the pytest method to check for exceptions: https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions

**Comment by wochinge** [comment]
@leon0707 
Sorry, I think communicated the location of the environment variables  not clearly. What I meant

```python
os.environ['USER_NAME'] = 'user'
os.environ['PASS'] = 'pass'

def test_read_yaml_string()
   [...]

# other tests [...]
```
So put it **above** `test_read_yaml_string` but not in the methods.

One more thing, then we are good to go: Can you maybe trim the example configuration within your test cases, so that it is not that much duplicate code.  I mean the actual values are not important, we just want to test the correct replacement of the environment variables. Just have the minimal required config for your test case. For example instead of:

```
    config_with_env_var = """
    tracker_store:
      mongo:
         host: http://localhost:27017
         db: rasa
         user: ${USER_NAME}
         password: ${PASS}
         collection: conversations
    """
```

Just:

```python
    config_with_env_var = """
    mongo:
         user: ${USER_NAME}
         password: ${PASS}
    """
```

File: tests/test_utils.py (hunk 6 above)
Comment (wochinge) [reply tests/test_utils.py]: @leon0707

File: CHANGELOG.rst [hunk 7]
```diff
@@ -37,6 +37,9 @@ Added
 - add documentation on how to configure endpoints within a configuration file
 - ``auth_source`` parameter in ``MongoTrackerStore`` defining the database to
   authenticate against
+- environment variables specified with ``${env_variable}`` in a yaml
+    configuration file are now replaced with the value of the environment variable
+
```
Comment (wochinge) [inline CHANGELOG.rst]:
```suggestion
```

**Pull request merged**